Add: orphan rule rationale.

TmSalviano opened the pull request:
1. I thought it would be helpful to introduce the reader to the problem the orphan rule solves before being exposed to the rule itself. Understanding the orphan rule for the first time can be challenging, and knowing why it exists beforehand can make it easier to grasp.
2. Also, including an explanation of the rationale behind the orphan rule in the orphan rule section of the page just makes sense.

TmSalviano committed:
Add: orphan rule rationale.

TmSalviano committed:
Fix:  grammar error . Fix:  improved semantics

Co-authored-by: Josh Triplett <josh@joshtriplett.org>

bluebear94 commented on src/items/implementations.md:
```diff
@@ -174,6 +174,15 @@ be instantiable with the same set of types for the input type parameters. -->
+implements the same trait for the same type, the compiler wouldn't know which implementation
+to use.\
+The orphan rule prevents this by requiring that either the trait or some type in the
+implementation is local to your crate, ensuring only one crate defines the implementation and
+thereby maintaining coherence.
```
I would personally put a “that” here:

```suggestion
implementation is local to your crate, ensuring that only one crate defines the implementation
and thereby maintaining coherence.
```

ehuss committed:
Update orphan rule introduction

This updates the orphan rule introduction based on some conversations
to try to emphasize a few points:

- Say what it is before explaining why it exists.
- Frame the rationale as saving space for crate authors.

ehuss reviewed (approved):
Thanks!

We had some conversations about reframing this a little, and so I pushed up some edits to try to capture some of the nuances that we discussed.
